### ISSUE TITLE
Cannot use analog models on multiple GPUs with torch.nnDataParallel 

### ISSUE THREAD

**Issue by jdbuhler** [IBM/aihwkit#578]
## Description

I am attempting to train an analog model (derived from a full-precision model) across multiple GPUs within one process, using PyTorch's nnDataParallel facility, roughly as follows:

model = AnalogSequential(
      convert_to_analog_mapped(fp_model, rpu_config)
)
model.remap_analog_weights()
pmodel = nn.DataParallel(model)
[go on to do training run using pmodel]

The rpu_config in question is an InferenceRPUConfig, and the model is from the PatchTST transformer family.

When I try to do this, I get the following error from the forward inference step in training:

  File "/gpfs/u/barn/AIPR/AIPRbhlr/PLP-multi/exp/exp_main.py", line 288, in _do_inference
    preds = model(inpts)
            ^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/parallel/data_parallel.py", line 185, in forward
    outputs = self.parallel_apply(replicas, inputs, module_kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/parallel/data_parallel.py", line 200, in parallel_apply
    return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/parallel/parallel_apply.py", line 110, in parallel_apply
    output.reraise()
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/_utils.py", line 694, in reraise
    raise exception
AttributeError: Caught AttributeError in replica 0 on device 0.
Original Traceback (most recent call last):
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/parallel/parallel_apply.py", line 85, in _worker
    output = module(*input, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/container.py", line 215, in forward
    input = module(input)
            ^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/barn/AIPR/AIPRbhlr/PLP-multi/models/PatchTST.py", line 90, in forward
    x = self.model(x)
        ^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/barn/AIPR/AIPRbhlr/PLP-multi/layers/PatchTST_backbone.py", line 74, in forward
    z = self.backbone(z)                                                                # z: [bs x nvars x d_model x patch_num]
        ^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/barn/AIPR/AIPRbhlr/PLP-multi/layers/PatchTST_backbone.py", line 162, in forward
    x = self.W_P(x)                                                          # x: [bs x nvars x patch_num x d_model]
        ^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/.local/lib/python3.11/site-packages/aihwkit-0.8.0-py3.11-linux-x86_64.egg/aihwkit/nn/modules/linear.py", line 97, in forward
    return self.analog_module(x_input)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/.local/lib/python3.11/site-packages/aihwkit-0.8.0-py3.11-linux-x86_64.egg/aihwkit/simulator/tiles/array.py", line 158, in forward
    result = analog_tile(x_input)
             ^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/.local/lib/python3.11/site-packages/aihwkit-0.8.0-py3.11-linux-x86_64.egg/aihwkit/simulator/tiles/inference.py", line 377, in forward
    out = AnalogFunction.apply(
          ^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/autograd/function.py", line 539, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/scratch/miniconda3x86/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs/u/home/AIPR/AIPRbhlr/.local/lib/python3.11/site-packages/aihwkit-0.8.0-py3.11-linux-x86_64.egg/aihwkit/simulator/tiles/functions.py", line 48, in forward
    use_indexed = analog_ctx.use_indexed
                  ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Tensor' object has no attribute 'use_indexed'

## How to reproduce

Attempt to train an analog model on > 1 GPU using nnDataParallel.

## Expected behavior

The model should not crash during forward evaluation.

## Other information

* Pytorch version: 2.1.0
* Package version: 0.8.0 (pulled and built on 12/20/2023
* OS: Linux, CUDA 11.8
* Python version: 3.11.5
* Conda version (or N/A): 23.11.0

**Comment by maljoras** [comment]
Hi @jdbuhler,
Note that you have to call `model.prepare_for_ddp()` to tell the ddp that it should exclude the syncing of the `analog_ctx`. Did you do that?

**Comment by jdbuhler** [comment]
I did not call prepare_for_ddp(), as I am not using DDP - this is nnDataParallel, which is multiple GPUs within one process.

Do I need to call prepare_for_ddp even in this case?

Jeremy

On Dec 22, 2023 13:27, "Malte J. Rasch" ***@***.***> wrote:

Hi @jdbuhler<https://github.com/jdbuhler>,
Note that you have to call model.prepare_for_ddp() to tell the ddp that it should exclude the syncing of the analog_ctx. Did you do that?

**Comment by kaoutar55** [comment]
@jdbuhler have you tried to the prepare function?

**Comment by maljoras** [comment]
Hi @jdbuhler ,
still you want exclude the syncing of the analog_ctx. This is done tagging these tensors as to be ignored for DDP. I never used `nn.DataParallel` but there should be a similar functionality to ignore these tensors.

**Comment by maljoras** [comment]
Also note that it is recommended to use DDP always (see [torch documentation](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html)): 
![image](https://github.com/IBM/aihwkit/assets/17587387/2d90f45a-c185-43bb-9519-3585898cec85)

**Comment by maljoras** [comment]
Closing this issue since DDP can be used instead of nnDataParallel.

**Comment by jdbuhler** [comment]
On 12/26/2023 11:44 AM, Kaoutar El Maghraoui wrote:
> @jdbuhler have you tried to the prepare function?

I added prepare_for_ddp(), but the error with nnDataParallel is unchanged.

I see that Malte has closed the GitHub issue.  It's OK if we need to use 
the DDP version exclusively, now that the issue has been addressed there 
(one less tree for me to maintain), but in that case, the AIHWKit docs 
should clearly state that nnDataParallel is not supported.  While DDP is 
the recommended way to use PyTorch, nnDataParallel is still a supported 
feature of the base package.

Jeremy
